### DISCUSSION TITLE
Nuxt SSG generation - Am I doing it right?

### DISCUSSION THREAD

**Discussion by PMendesWebDev** [nuxt/nuxt#22585] [Questions]
Hello there!

I've a nuxt app that is supposed to be built into a static site using "npx nuxi generate". It uses the .vue pages in the /pages folder.

I have a couple of issue here:

**1.** Isn't this suppose to render a static site (plain html, js, css)? Why is it creating a _nuxt folder with all sort of files inside - like _component_.vue._number_.js; index._number_.js (so many of this ones)? Isn't the content in the html file already?

**2.** No matter what I did in the nuxt.config.ts file, the only way I got this to work (prerendering all routes) was to to this:

```
nitro: {
    prerender: {
      crawlLinks: false,
      routes: getPages(process.env.BUILD_SECTION ?? ''),
      ignore: ['/dev'],
    },
  },
```

where the getPages() is reading the /pages dir to make sure all routes are there

```
function getPages(dir: string): string[] {
  let pages: string[] = []

  let content = fs.readdirSync(`./pages/${dir}`)

  if (content && content.length) {
    for (let i = 0; i < content.length; i++) {
      const currentItem = `${dir}/${content[i]}`
      if (currentItem.endsWith('.vue')) {
        continue
      }
      pages.push(currentItem)

      pages = pages.concat(getPages(currentItem))
    }
  }

  return pages
}
```

(if I didn't do this, the only rendered files would be 200.html and 404.html

**3.** I have two .vue files in the /pages root that I want to prerender but if I add them like so in the prerender.routes array (i.e. 200.vue, 400.vue) it throws an exception and fails

I'll appreciate any and every help you could give.

Thanks,
Pedro

**Comment by TheAlexLichter** [comment]
1. Static site which will transform in an SPA after the first request. If you don't need JS, you can fully remove JS too (e.g. https://nuxt.new/ is doing that)
2. If you have static sites and all your sites are linked properly via anchor tags, you don't need  *any* config. `nuxi generate` should be sufficient. It will pick up all "static files" (like `about.vue`) up, render and crawl the rendered pages for more links. That way, it should also covered something like `post/[id].vue`, as long as all the posts are linked somewhere

**Comment by PMendesWebDev** [comment]
"Static site which will transform in an SPA after the first request. If you don't need JS, you can fully remove JS too (e.g. https://nuxt.new/ is doing that)" 
- isn't that SSR? static site should be a static site, no SPAs whatsoever. I'd expect the output of a generate to be plain html, css and js files. No nuxt routers, no nuxt loader, no nuxt at all

**Reply by TheAlexLichter** [reply]
> isn't that SSR?

It is "SSR at build time".

> static site should be a static site, no SPAs whatsoever. 

The "static" in "static site" means **static assets**. And that is the case after running `nuxi generate`

> I'd expect the output of a generate to be plain html, css and js files.

That's the case

> No nuxt routers, no nuxt loader, no nuxt at all

I don't fully understand this part though. You want JS, but not Nuxt/Vue "behind it"? 🤔 

PS: Please use the "reply" function in discussion below the post so it is easier to follow along ☺️ 
![grafik](https://github.com/nuxt/nuxt/assets/640208/60afd5b2-19bc-4691-b496-ab4a244b1fea)

**Reply by PMendesWebDev** [reply]
Yeah, I think I've totally misunderstood SSG.

SSG is pretty much SSR but all the routes are prerendered, right? SSR will render the requested route and SSG will have it rendered already, they both act as a SPA after the first request, isn't it?

I believe we have made a huge mistake in the setup of this project.

We have an application that creates .vue files (`/pages` and /`layouts`) based on a trigger in the CMS. The files are commited to a repo that runs `npx nuxi generate` and deploys it to hosting (we are hosting - due to several reasons - the generated site in azure web app - not static web app - and it will run in IIS).

Multilingual is needed so we have .vue inside `/laouts` and .vue inside `/pages` for each language. The `/pages` structure is something like this:

```
/layouts
|-  layoutPT.vue
|   layoutEN.vue

/pages
|-  pt
|   |- index.vue
|   |- other_folder_pt
|   |  
|   ...
|-  en
|   |- index.vue
|   |- other_folder_en
|-  index.vue
```

the `/pages/index.vue` is the same as `/pages/pt/index.vue` (with canonical sets and all that)

I was hoping to get IIS's error handling (in web.config) to show page404 with layoutPT by default and page404 with layoutEN wherever we are inside `/en`, but nuxt routing is preventing that from happening by default (since the SPA kicks in after the first load). This means that 404 erros that come from in-app navigation not show the custom 404 page but 404 coming from direct URL access or page reloads work.

And since we are upgrading a client's website, redirects must be made (from old site url to new site url) and most of them don't follow a rule that could be handled with regex, they will be hardcoded...


Regarding your last point (since the previous one were me being dumb):

> > No nuxt routers, no nuxt loader, no nuxt at all

> I don't fully understand this part though. You want JS, but not Nuxt/Vue "behind it"? 

Since I mistakenly thought that SSG would only need nuxt for the build time, I wasn't expecting nuxt application to be needed at "run" time...

Oh man, oh man, oh man... This is what you get for not grasping at least 70% of the framework you are going to work with..

**Reply by TheAlexLichter** [reply]
> SSG is pretty much SSR but all the routes are prerendered, right?

Correct! That's why SSG is also a synonym for Prerendering 😋 

> SSR will render the requested route and SSG will have it rendered already, they both act as a SPA after the first request, isn't it?

Also correct!

> This means that 404 erros that come from in-app navigation not show the custom 404 page but 404 coming from direct URL access or page reloads work.

By default I'd say so, yes. But you could alter your logic so the 404 from internal navigation would do a `window.location.reload()` to show your IIS error page. *Or* write a custom error page in Nuxt and show this all the time (+ configure IIS to serve that). Both is feasible 😋

**Reply by PMendesWebDev** [reply]
> By default I'd say so, yes. But you could alter your logic so the 404 from internal navigation would do a window.location.reload() to show your IIS error page. Or write a custom error page in Nuxt and show this all the time (+ configure IIS to serve that). Both is feasible 😋

This seems a solution that would solve all my problems!

Do you happen to know how I could do a window.location.reload() on a 404 error?

**Reply by TheAlexLichter** [reply]
@PMendesWebDev Hmm. You could try that in your `error.vue`. But don't forget to check the error status code ☺️

**Comment by colinlienard** [comment]
Have you set `ssr: false` in your nuxt config? I believe this will block prerendering as mentionned [here](https://nuxt.com/docs/api/configuration/nuxt-config#ssr).

Also, the `.nuxt` folder is not the output folder, `dist` and `.output` are.

**Reply by PMendesWebDev** [reply]
> Have you set `ssr: false` in your nuxt config? I believe this will block prerendering as mentionned [here](https://nuxt.com/docs/api/configuration/nuxt-config#ssr).

I have `ssr: true` in my nuxt.confg

> Also, the `.nuxt` folder is not the output folder, `dist` and `.output` are.

I was talking about the _nuxt folder that is created inside the `.output/public` folder :)

**Reply by kai713** [reply]
thanks man, u helped me
